### PR TITLE
embeddings: Fetch archive vs thousands of RPC calls

### PR DESCRIPTION
**Pull request by eseliger**
This PR moves away from a model that looks at each individual file in favor of fetching a tar archive of the worktree and iterating over it locally from the tar file.
This PR makes worker require some TMPDIR storage to be available and writable.
Sorry that this diff is a bit ugly, there's definitely room for making this nicer to read, but I've tried to keep the diff as contained as possible so it is easier to backport. We can revamp this more a bit later.

## Test plan

Verified manually things still index and adjusted tests.

### CODE REVIEW

**Comment by eseliger** [comment]
we usually say that bandwidth is fairly unlimited between services in the same VPC/same machine, and gzipping adds CPU overhead, so it's likely not super worth it. I don't have numbers to point to, but that's at least the reasoning to my understanding why we also don't do it on the searcher tar fetching path. 
It's all HTTP btw since we don't do HTTPS for inter-service communication

**Review by vdavid — COMMENTED** [review]
Nice cleanup and optimization! It's not my domain so not approving, but the change LGTM!

File: enterprise/cmd/worker/internal/embeddings/repo/handler.go [hunk 1]
```diff
@@ -83,18 +79,132 @@ func (h *handler) Handle(ctx context.Context, logger log.Logger, record *repoemb
...
+					}
+					if thr.Size < embed.MIN_EMBEDDABLE_FILE_SIZE {
+						continue
+					}
+					// Skip files that are excluded.
+					if embed.IsExcludedFilePath(thr.Name, excludedGlobPatterns) {
+						continue
+					}
+					// Only emit text files.
+					if !embed.IsValidTextFile(thr.Name) {
```
Comment (novoselrok) [inline enterprise/cmd/worker/internal/embeddings/repo/handler.go:157]: Is the negated `embed.IsValidTextFile(thr.Name)` the only change from the function above? Could the `filesReader` function also accept a boolean flag (e.g., `onlyCodeFiles`) and switch between the two cases?

File: enterprise/cmd/worker/internal/embeddings/repo/handler.go [hunk 2]
```diff
@@ -83,18 +79,132 @@ func (h *handler) Handle(ctx context.Context, logger log.Logger, record *repoemb
...
 		return err
 	}
 
 	return embeddings.UploadRepoEmbeddingIndex(ctx, h.uploadStore, string(embeddings.GetRepoEmbeddingIndexName(repo.Name)), repoEmbeddingIndex)
 }
+
+// fetchTarToDisk retrieves a tar for the given repo at the given commit and writes
+// it to a temporary file.
+func fetchTarToDisk(ctx context.Context, gitserverClient gitserver.Client, repoName api.RepoName, commit api.CommitID) (*os.File, error) {
+	r, err := gitserverClient.ArchiveReader(ctx, authz.DefaultSubRepoPermsChecker, repoName, gitserver.ArchiveOptions{
```
Comment (novoselrok) [inline enterprise/cmd/worker/internal/embeddings/repo/handler.go:180]: Including `authz.DefaultSubRepoPermsChecker` will disable indexing repos with sub-repo permissions. https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/gitserver/commands.go?L2386:28

We check sub-repo permissions at search time and filter out files then.

Comment (eseliger) [reply enterprise/cmd/worker/internal/embeddings/repo/handler.go:180]: We use an internal actor here and internal actors are to my knowledge exempt from permissions enforcement.

File: enterprise/cmd/worker/internal/embeddings/repo/handler.go (hunk 1 above)
Comment (eseliger) [reply enterprise/cmd/worker/internal/embeddings/repo/handler.go:157]: yeah, I'll do that.

Comment (eseliger) [reply enterprise/cmd/worker/internal/embeddings/repo/handler.go:157]: done

**Comment by eseliger** [comment]
@novoselrok I think I addressed the feedback :)

**Comment by eseliger** [comment]
Encountered a blocker for this, we cannot write to disk in the worker pod at the moment. I think we should allow that, but getting it into the various deployment methods might take time so I don't think this'll be feasible for 5.0.1 at this point 😞 
Without writing the tar to disk, we need to keep a connection to gitserver open for potentially hours, which is gonna cause other problems ..

**Comment by peterguy** [comment]
> we cannot write to disk in the worker pod at the moment

can we extract to memory? Or do you have to write to disk because of `ReadCloser`?

**Comment by eseliger** [comment]
We would like to avoid requiring an additional ~4GB of memory on the worker just to keep the repo archive 😞 Potentially more if the repo is even bigger than our megarepo.

**Review by keegancsmith — COMMENTED** [review]
ok from reviewing the code, the suggested approach I mentioned before actually matches more closely how this worked before. It might actually be quite minimal to add:

- use a git ls-tree call which includes object sizes instead of the older code which just got filenames then used stat.
- Fetch files by object ID.
- (Possibly optional) Improve perf by batching fetches of files by object ID.

**Comment by keegancsmith** [comment]
> That's interesting, we moved away from this approach because it required a lot of RPC calls to read the files, and memory to keep all files of a worktree, such as from ls-files. In the proposed approach here, we require only little memory to read the tar header, and then a bit later the file which is capped to 1MB. Also, it makes the process more resilient to gitserver outages.
> The stat call I see we could get rid of with that approach though.
> Am I missing something about the memory consumption of keeping potentially tens of thousands of file paths in memory here? 🤔

Ok, I underestimated the size for all the file paths. The size of the output for ls-tree that I propose on megarepo is about 50MB. Given we have customers with monorepos bigger than this, it does indeed feel like a non-starter. Time to start thinking about interesting ways to paginate / reduce working set sizes. But yeah, if it wasn't for that size being so big, then the ls-tree + batch git cat-file approach is pretty much the same performance as git archive except we have much more control.

```
gitserver-8:/data/repos/github.com/sgtest/megarepo# git ls-tree -r -l -t -z HEAD | wc -c
49731551
```

**Pull request closed without merging**